### PR TITLE
[AOT] Add AOTLowerMain pass to lower a Relay main into TIR

### PR DESCRIPTION
**Pull request by mbaret**
Tracking issue: https://github.com/apache/tvm/issues/12548

This is a pass refactored out of the AOTExecutorCodegen. Instead of combining all of the functionality of the AOTExecutorCodegen into a single monolithic pass, this pass only handles the lowering of the Relay main function into TIR. Tests for the pass are included.

cc @Mousius @alanmacd @areusch

### CODE REVIEW

**Comment by mbaret** [comment]
cc @gigiblender @mikepapadim @leandron @ekalda

**Review by Mousius — CHANGES_REQUESTED** [review]
Nice work @mbaret! I'm assuming this is mostly a move and a clean up with additional test coverage? Based on that I think this is already much better than what we have but it'd be good to split out the `ExprAllocator` pass and test it independently - otherwise I think we can get this in to move forwards with the larger refactor

**Comment by mbaret** [comment]
> It'd be good to split out the `ExprAllocator` pass and test it independently

I did consider this, but sided against it on the basis that `ExprAllocator` is an implementation detail of `AOTMainLowerer`. We should still be able to test all the behaviours implemented by `ExprAllocator` through the stable interface of `AOTMainLowerer`.

I accept there are some conflicting testing philosophies around this though (I think BDD vs. TDD being the relevant one here), and so am open to other views.

File: tests/python/relay/aot/test_pass_aot_lower_main.py [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+
+
+def transform(mod):
+    host_target = tvm.target.Target("llvm")
+    prim_target = tvm.target.Target("llvm", host=host_target)
+    ctxt = tvm.transform.PassContext()
+    config = tvm.target.make_compilation_config(ctxt, prim_target)
+    mod = tvm.relay.transform.PlanDevices(config)(mod)
+    mod = tvm.relay.transform.InferType()(mod)
+    if mod.get_attr("device_contexts") is None:
```
[the remark this answers is not in the document]
Comment (mbaret) [reply tests/python/relay/aot/test_pass_aot_lower_main.py:42]: The idea with this is that if you add it in a test it won't be overwritten, but if you don't add it then it'll get set to a sensible default. I agree hiding this logic in a helper func reduces test clarity though, so will look into a more explicit function (or see below...)

side note: I really hate these magic attributes in the compiler and they make tests unnecessarily hard to follow by essentially 'cheating' the IRModule->IRModule property. I'm going to take another look at the refactored pass to see if either I can make `device_contexts` an explicit argument to initialize the pass, or derive it from the IRModule directly.

File: tests/python/relay/aot/test_pass_aot_lower_main.py [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+
+def make_const(dtype, shape):
+    return tvm.relay.const(np.zeros(shape).astype(dtype))
+
+
+def make_consts(dtype, shapes):
+    return [make_const(dtype, shape) for shape in shapes]
+
+
+def transform(mod):
```
[the remark this answers is not in the document]
Comment (mbaret) [reply tests/python/relay/aot/test_pass_aot_lower_main.py:35]: agree with the sentiment, but maybe `_plan_devices` so we don't overload the concept of lowering main?

[the remark this answers is not in the document]
Comment (Mousius) [reply tests/python/relay/aot/test_pass_aot_lower_main.py:35]: Ah yeah, that would be better, I think my head thought this one did the lowering bit as well 🤔

File: tests/python/relay/aot/test_pass_aot_lower_main.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Mousius) [reply tests/python/relay/aot/test_pass_aot_lower_main.py:42]: The second option is clearer but the information will still need to be threaded from `LowerTE` to `LowerMain` due to the first pass producing a new `IRModule` without any way of providing the device information (which makes sense given the `IRModule` is meant to contain all this information). Maybe you could also re-scan the `main` function within AOT to re-create the device contexts rather than using attributes?

[the remark this answers is not in the document]
Comment (mbaret) [reply tests/python/relay/aot/test_pass_aot_lower_main.py:42]: I found your TODO for this haha https://github.com/apache/tvm/blob/a0fe74b3c3608929b21faeaea422ac09aa2f75eb/src/relay/backend/te_compiler.cc#L1230-L1231

It looks like some destructive lowering takes place in `LowerTE` removing `kCompiler` attrs that are otherwise saved into this data structure. One option would be to have a 'device_context' attr defined for each PrimFunc (which I consider a lesser sin that IRModule level attrs).

**Comment by Mousius** [comment]
> > I'm assuming this is mostly a move and a clean up with additional test coverage?
> 
> Yeah that's right, we were finding the the AOT codegen had become unmanageably complex and had very poor testing coverage so it was difficult to add features to. This is a first attempt at bringing it back from the brink :) There's some more context on where this is headed in the tracking issue.
> 

Makes sense, I reviewed it in this spirit rather than a complete rewrite 😸 

> > It'd be good to split out the `ExprAllocator` pass and test it independently
> 
> I did consider this, but sided against it on the basis that `ExprAllocator` is an implementation detail of `AOTMainLowerer`. We should still be able to test all the behaviours implemented by `ExprAllocator` through the stable interface of `AOTMainLowerer`.
> 
> I accept there are some conflicting testing philosophies around this though (I think BDD vs. TDD being the relevant one here), and so am open to other views.

In either the BDD or the TDD case you'd have tests first so no code would appear without a test to cover it. I don't think that's the case here as the tests are retro-fitted to the existing code?  If written tests first, I would expect to follow the method you've described by starting with tests for `AOTMainLowerer` and then write tests for the internal units where I can't test them as part of the higher level interface. 

I don't think you can test all the behaviours of `ExprAllocator` through `AOTMainLowerer` due to cases such as:
https://github.com/apache/tvm/pull/12550/files#diff-fd79a3a236f19a9f85e5045e7c75ed47a09ad7d698042114a93f4f74dd17c75eR99-R101
Which is an internal optimisation inside of `ExprAllocator` that hopefully we don't remove but makes no difference to the outer behaviour of the AOT pass and is therefore difficult to assert is behaving correctly.

The [TVM Code Quality Guidelines](https://github.com/apache/tvm/blob/main/docs/contribute/code_review.rst#factors-to-consider-about-code-quality) philosophy is that we should ensure test coverage, implying we should test these cases and also error cases which are equally part of the `Pass`es contract.

File: tests/python/relay/aot/test_pass_aot_lower_main.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Mousius) [reply tests/python/relay/aot/test_pass_aot_lower_main.py:42]: Oh no 🙀  I'm now curious why past-@Mousius didn't attach it as `kTarget` to the `PrimFunc` which leads me to think that there was some other logic that broke due to that 🤔 I think for now attaching a device context instead will be less dangerous.

**Comment by mbaret** [comment]
> Which is an internal optimisation inside of ExprAllocator that hopefully we don't remove but makes no difference to the outer behaviour of the AOT pass and is therefore difficult to assert is behaving correctly.

If that code is unreachable from AOTMainLowerer, it is fundamentally unreachable from the program that is 'TVM', and at the end of the day TVM is the component under test. If it is reachable, then we can test it through the AOTMainLowerer interface.

I think it is reachable because we can always construct a test case with a function marked `kPrimitive` for AOTMainLowerer. If that's the primary concern here I'll happily add in such a test. I would caution though that the [TVM Code Quality Guidelines](https://github.com/apache/tvm/blob/main/docs/contribute/code_review.rst#factors-to-consider-about-code-quality) make no commitment to 100% coverage, so I think it's important we focus our testing effort wisely (I certainly haven't covered every corner case in the current tests).

File: tests/python/relay/aot/test_pass_aot_lower_main.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Mousius) [reply tests/python/relay/aot/test_pass_aot_lower_main.py:42]: Also, back to my initial question, doesn't this mean that an empty `device_contexts` attr is totally valid for any `IRModule` without a `kCompiler` annotated function which also has a `Target` with `use_device_api` ?

**Comment by Mousius** [comment]
> > Which is an internal optimisation inside of ExprAllocator that hopefully we don't remove but makes no difference to the outer behaviour of the AOT pass and is therefore difficult to assert is behaving correctly.
> 
> If that code is unreachable from AOTMainLowerer, it is fundamentally unreachable from the program that is 'TVM', and at the end of the day TVM is the component under test. If it is reachable, then we can test it through the AOTMainLowerer interface.

The line itself is not unreachable, it's harder to test and ensure the behaviour from within `AOTMainLowerer` as you're essentially testing "nothing happened" with the `kPrimitive` attribute added. There are always pieces of behaviour which are not explicitly user facing but are important for the overall behaviour of the program, else you could consider the existing AOT tests which invoke `relay.build` to be sufficient?
 
> I think it is reachable because we can always construct a test case with a function marked `kPrimitive` for AOTMainLowerer.  If that's the primary concern here I'll happily add in such a test. I would caution though that the [TVM Code Quality Guidelines](https://github.com/apache/tvm/blob/main/docs/contribute/code_review.rst#factors-to-consider-about-code-quality) make no commitment to 100% coverage, so I think it's important we focus our testing effort wisely (I certainly haven't covered every corner case in the current tests).

Yip, you can interpret "test coverage" differently, I'd rather not gamble on whether a future change will be caught or not by the tests though 😸

**Comment by mbaret** [comment]
At the end of the day, I think we're debating whether or not to test internal optimizations for what is essentially performance benefit. I'm especially wary to set precedent here because this very refactor would have been made a lot harder if extensive testing of internal implementation existed.

I appreciate this is effectively down to personal preference given no strict ruling in the code quality guidelines around acceptable coverage, and that makes discussions like this hard to progress. In the spirit of [Consensus Building](https://github.com/apache/tvm/blob/main/docs/contribute/code_review.rst#consensus-building), perhaps we can wait and see if there's any other input given I think we've both presented the extent of our arguments at this stage.

File: python/tvm/relay/backend/utils.py [hunk 3]
```diff
@@ -15,6 +15,13 @@
 # specific language governing permissions and limitations
 # under the License.
 """Utility backend functions."""
+from enum import Enum
+
+
+class CallType(Enum):
```
Comment (lhutton1) [inline python/tvm/relay/backend/utils.py:21]: just speculating: is it possible to get these values from the enum in c++ in any way? Just thinking there's the possibility that changes to the enum in c++ might not be reflected here

File: src/relay/backend/aot/aot_lower_main.cc [hunk 4]
```diff
@@ -0,0 +1,869 @@
...
+/*!
+ * \brief Assigns one or more StorageInfos to expressions requiring storage in a
+ * function to produce an Expr to StorageInfo map.
+ *
+ * This pass is leveraged by AOTMainLowerer to perform an initial naive allocation
+ * for tensors in the Relay main function. The resulting storage map is then lowered
+ * into TIR allocations by AOTMainLowerer where the allocation can be subsequently
+ * optimized by later passes (e.g. USMP).
+ */
+class ExprAllocator : public transform::DeviceAwareExprVisitor {
```
Comment (lhutton1) [inline src/relay/backend/aot/aot_lower_main.cc:54]: In general I'd say its easier (at least for me) to keep track of objects like `StorageInfos` as attributes on the IR, rather than doing what is done here by maintaining the `StorageMap` as a sort of side-channel. Likely out of scope for this PR but just wondering what your thoughts are?

**Review by lhutton1 — COMMENTED** [review]
As someone who recently tried to make sense of the AOT codegen this is definitely a welcomed change :) I see that nothing was removed from `aot_executor_codegen.cc`, I'm assuming that this will come in a later PR?

File: src/relay/backend/aot/aot_lower_main.cc (hunk 4 above)
Comment (mbaret) [reply src/relay/backend/aot/aot_lower_main.cc:54]: I think in this case I'd rather consider it an implementation detail rather than a side-channel - that is to say we're not channeling any information alongside a pass, it gets consumed immediately as part of AOTMainLowerer.

File: python/tvm/relay/backend/utils.py (hunk 3 above)
Comment (mbaret) [reply python/tvm/relay/backend/utils.py:21]: If it is I'm not sure how :P Agreed there's a drifting risk although hopefully the tests would detect this.

**Comment by mbaret** [comment]
> I see that nothing was removed from aot_executor_codegen.cc, I'm assuming that this will come in a later PR?

Yeah, aot_executor_codegen.cc only gets refactored itself in P4. The first 3 PRs are to extract functionality from aot_executor_codegen.cc as standalone passes, these then get combined to build the newly refactored codegen. I've got a draft PR demonstrating this if you want to see how that ends up looking: https://github.com/apache/tvm/pull/12549.

File: python/tvm/relay/backend/utils.py (hunk 3 above)
Comment (lhutton1) [reply python/tvm/relay/backend/utils.py:21]: Yeah I didn't have anything in mind either :)

**Comment by areusch** [comment]
>>I think it is reachable because we can always construct a test case with a function marked kPrimitive for AOTMainLowerer. If that's the primary concern here I'll happily add in such a test. I would caution though that the [TVM Code Quality Guidelines](https://github.com/apache/tvm/blob/main/docs/contribute/code_review.rst#factors-to-consider-about-code-quality) make no commitment to 100% coverage, so I think it's important we focus our testing effort wisely (I certainly haven't covered every corner case in the current tests).
>
> Yip, you can interpret "test coverage" differently, I'd rather not gamble on whether a future change will be caught or not by the tests though 😸

@Mousius regarding test coverage: generally I'm supportive of having more test coverage, but i'm not sure in this case it's necessary to block the PR over it; in this case, i think the regression is that we'd visit more of the main function than we'd need to in building `expr_storage_map_`. to me this falls a bit more on the "perfect" side of the don't-let-perfect-be-the-enemy-of-good argument. is there another kind of regression you can see here that is hard to test without explicitly testing ExprAllocator standalone?

**Review by areusch — COMMENTED** [review]
one other minor comment here

File: src/relay/backend/aot/aot_lower_main.cc [hunk 5]
```diff
@@ -0,0 +1,869 @@
...
+#include "../../op/memory/device_copy.h"
+#include "../../op/memory/memory.h"
+#include "../../transforms/device_aware_visitors.h"
+#include "../name_transforms.h"
+#include "../utils.h"
+
+namespace tvm {
+namespace relay {
+namespace backend {
+namespace executor {
```
[the remark this answers is not in the document]
Comment (mbaret) [reply src/relay/backend/aot/aot_lower_main.cc:39]: Fair comment, early on I thought of adding the executor namespace because 'backend' seemed to have quite a wide variety of stuff in it. Reflecting on this after doing most of the refactor, I agree it's probably not needed (I never got any value out of it at least).

**Comment by mbaret** [comment]
I'm keen to make progress with this issue and given the AOT codegen is largely untested, I believe this refactor represents a strict improvement over the current situation. The decisions I've made here aren't meant to be final and patches are always welcome to introduce further coverage, including adding specific test cases where community stakeholders care about particular functionality not regressing.

To give a more blunt personal view, there's an enormous amount of untested code in TVM with a lot of it poorly specified. If we make the barrier-to-entry for participating in improving that situation '100% coverage', rather than getting full coverage we'll just get no participation. That being said, if we want to update the project's code quality guidelines to be more explicit around testing expectations, I'd be fully supportive of that effort and would happily participate in any discussions arising from an RFC. We can then vote as a community on how we want to go forward rather than having the same debates on a PR-by-PR basis.

@lhutton1 could you request further changes/approve, thanks :)
@Mousius let me know what you think

**Review by areusch — APPROVED** [review]
thanks @mbaret ! @Mousius could you take a look?

File: src/relay/backend/aot/aot_lower_main.cc [hunk 6]
```diff
@@ -0,0 +1,868 @@
...
+namespace relay {
+namespace backend {
+namespace aot {
+
+using StorageMap =
+    std::unordered_map<Expr, StorageInfo, runtime::ObjectPtrHash, runtime::ObjectPtrEqual>;
+
+/*!
+ * \brief Assigns one or more StorageInfos to expressions requiring storage in a
+ * function to produce an Expr to StorageInfo map.
```
Comment (ekalda) [inline src/relay/backend/aot/aot_lower_main.cc:46]: Would you mind making this sentence a bit clearer? It only started to make sense to me after I had looked at the code.

File: src/relay/backend/aot/aot_lower_main.cc [hunk 7]
```diff
@@ -0,0 +1,868 @@
...
+
+    // Assign StorageInfo to all the Relay exprs
+    ExprAllocator expr_allocator;
+    expr_allocator.Run(lowered_main_func);
+    expr_storage_map_ = expr_allocator.GetStorageMap();
+
+    for (auto input : lowered_main_func->params) {
+      input_vars_.push_back(input);
+      std::string input_name = SanitizeName(input->name_hint());
+      // We dont want the compiler changing input names in the
```
Comment (ekalda) [inline src/relay/backend/aot/aot_lower_main.cc:237]: Nit: I know it is copy pasted code, but
```suggestion
      // We don't want the compiler changing input names in the
```

File: src/relay/backend/aot/aot_lower_main.cc [hunk 8]
```diff
@@ -0,0 +1,868 @@
...
+      } else {
+        CreateIOVar(lowered_main_func->body, "output", /*use_unique_name = */ false);
+      }
+    }
+
+    CollectDeviceVariables(lowered_mod->GetAttr<Map<GlobalVar, String>>("device_contexts")
+                               .value_or(Map<GlobalVar, String>()));
+    VisitExpr(lowered_main_func->body);
+
+    lowered_mod->Remove(lowered_mod->GetGlobalVar("main"));
```
Comment (ekalda) [inline src/relay/backend/aot/aot_lower_main.cc:291]: Nit: There used to be a comment there about removing Relay main and replacing it with TIR one which I think was quite useful.

File: src/relay/backend/aot/aot_lower_main.cc [hunk 9]
```diff
@@ -0,0 +1,868 @@
...
+    auto tir_main_func = CreateMainFunc(mod_name);
+    lowered_mod->Update(GlobalVar(runtime::symbol::tvm_module_main), tir_main_func);
+    lowered_mod = tir::transform::RemoveNoOp()(lowered_mod);
```
Comment (ekalda) [inline src/relay/backend/aot/aot_lower_main.cc:294]: What's the reason for running this pass here instead of the "AoT pipeleline"?

**Review by ekalda — COMMENTED** [review]
Sorry for being late to the party - just some nits :) In general, I'm really happy to see this bit of codebase being cleaned up, so thanks a lot @mbaret ! 

Regarding to testing the `ExprAllocator`, what's the problem with testing it? Would it be very difficult to do? It has quite a bit of logic in it (and at least to me, everything that has visitors becomes somewhat unpredictable and hard to reason about), so my instincts align here with @Mousius. But if the gains from testing it are marginal compared to the effort, I'd rather not stop that PR going ahead because of this.

File: src/relay/backend/aot/aot_lower_main.cc (hunk 9 above)
Comment (mbaret) [reply src/relay/backend/aot/aot_lower_main.cc:294]: Otherwise the output of `AOTLowerMain` has a bunch of empty statements in it which I think would be very confusing behaviour to test.

**Comment by Mousius** [comment]
> > > I think it is reachable because we can always construct a test case with a function marked kPrimitive for AOTMainLowerer. If that's the primary concern here I'll happily add in such a test. I would caution though that the [TVM Code Quality Guidelines](https://github.com/apache/tvm/blob/main/docs/contribute/code_review.rst#factors-to-consider-about-code-quality) make no commitment to 100% coverage, so I think it's important we focus our testing effort wisely (I certainly haven't covered every corner case in the current tests).
> > 
> > 
> > Yip, you can interpret "test coverage" differently, I'd rather not gamble on whether a future change will be caught or not by the tests though 😸
> 
> @Mousius regarding test coverage: generally I'm supportive of having more test coverage, but i'm not sure in this case it's necessary to block the PR over it; in this case, i think the regression is that we'd visit more of the main function than we'd need to in building `expr_storage_map_`. to me this falls a bit more on the "perfect" side of the don't-let-perfect-be-the-enemy-of-good argument. is there another kind of regression you can see here that is hard to test without explicitly testing ExprAllocator standalone?

I've requested a minimal amount of additional tests be added to this PR, which is attempting to address a [broken window](https://en.wikipedia.org/wiki/Broken_windows_theory) in the code base. Please don't get hung up on the singular example I found from a brief review, what troubles me more is the active resistance to investing effort in testing, this is not good and far from perfect. If we are aware the behaviour could regress, and we can solidify the behaviour with a test case, I don't understand the rationale for ignoring it? We're already ignoring a large amount of the overall logic by not testing error cases, which in itself means we can't guarantee the component fails safely.

**Comment by mbaret** [comment]
To be clear, this isn't just a case of me refusing to put the test in 'just because it's effort' - I don't agree at a fundamental level with testing an implementation detail that has no _functional_ impact on the behaviour of the component-under-test. Because of this, I don't want to establish precedent for what I believe to be bad practice (such tests have created nightmares for me in the past when refactoring). I accept you have a different view, and I've outlined a path by which the community can come to a consensus on testing best-practices. I don't think the comments on a PR with incredibly limited visibility is the forum for this though.

File: src/relay/backend/aot/aot_lower_main.cc (hunk 6 above)
Comment (mbaret) [reply src/relay/backend/aot/aot_lower_main.cc:46]: Yeah can do, roughly what would you like to see here? I think I'm now over-familiar with the code so good to have fresh eyes.

**Comment by mbaret** [comment]
cc @ekalda @lhutton1 could you take another look and see if you've got further comments or can approve?

**Comment by mbaret** [comment]
@Mousius As this PR is being blocked and none of my proposals to achieve [consensus](https://tvm.apache.org/docs/contribute/code_review.html#consensus-building) have been acknowleged, I have done the following:

- A new function, `CreateStorage`, has been created and exposed in the header to allow `ExprAllocator` to be subjected to testing.

- A test has been added to confirm that `StorageInfo` is not allocated for nested functions as requested.

Because ExprAllocator is coupled to the implementation of `AOTLowerMain`, I've added a comment to indicate this in the header. Unfortunately, there was no programmatic way to enforce that this function isn't reused outside `AOTLowerMain` without making it untestable. Furthermore, the logic under test is unreachable from `AOTLowerMain` as AOT does not support nested functions and will check fail if they're encountered.

Lastly, I would prefer going forward that if committers intend to block PRs indefinitely, in parallel they engage with a consensus building process involving multiple members of the community. Without such engagement, contributors are disempowered to take part in good faith technical discussions which is a [core](https://tvm.apache.org/docs/contribute/code_review.html#building-trust) [tenant](https://tvm.apache.org/docs/contribute/code_review.html#community-participation) of the [review process](https://tvm.apache.org/docs/contribute/community.html#general-development-process).

@lhutton1 @ekalda please [approve or request changes explicitly](https://tvm.apache.org/docs/contribute/code_review.html#approve-and-request-changes-explicitly).

**Review by lhutton1 — APPROVED** [review]
Many apologies for the delay in taking another look at this - thanks @mbaret LGTM with regards to my comments. With regards to the testing, I don't really hold a strong opinion so will leave for others to take a look

**Review by Mousius — CHANGES_REQUESTED** [review]
> @Mousius As this PR is being blocked and none of my proposals to achieve [consensus](https://tvm.apache.org/docs/contribute/code_review.html#consensus-building) have been acknowleged

It seems to me there was an active discussion on this PR with various interested parties participating which leads me to believe we are working towards consensus? The code review guidelines were also contributed via voting thread and explicitly state test coverage is expected for code contributions. 

> Because ExprAllocator is coupled to the implementation of `AOTLowerMain`, I've added a comment to indicate this in the header. Unfortunately, there was no programmatic way to enforce that this function isn't reused outside `AOTLowerMain` without making it untestable. Furthermore, the logic under test is unreachable from `AOTLowerMain` as AOT does not support nested functions and will check fail if they're encountered.

I think this is a great outcome of a code review, by attempting to find an appropriate way of testing this I believe we've found this code is actually unreachable and can be removed? This is a good reason to encourage more discussions to improve test coverage, as it helps to highlight code which is unreachable (or requiring a slightly more focussed test as was first thought). In this case, can we just remove the unreachable code? If it's re-added I would expect we can re-add with the appropriate tests.

> Lastly, I would prefer going forward that if committers intend to block PRs indefinitely, in parallel they engage with a consensus building process involving multiple members of the community. Without such engagement, contributors are disempowered to take part in good faith technical discussions which is a [core](https://tvm.apache.org/docs/contribute/code_review.html#building-trust) [tenant](https://tvm.apache.org/docs/contribute/code_review.html#community-participation) of the [review process](https://tvm.apache.org/docs/contribute/community.html#general-development-process).

It was never my intention to block this PR indefinitely, I raised my request for changes early and have been active in the technical discussion to work with you to move this forwards. As noted, the code review guidelines were voted in by the community, if you wish to make a change to reduce test coverage, I propose you raise a process RFC which we can discuss and vote on. Contributors should be aware that "High-quality code reviews prevent technical debt for long-term and are crucial to the success of the project", in which case there are likely to be discussions and changes requested - I'd hope a contributor feels empowered by an active review thread and welcome feedback on their contributions?

**Comment by ekalda** [comment]
Apologies, I was away last week, so just caught up with the discussion here. 

@mbaret I'm a bit confused by the addition of `CreateStorage` - you added handling (and testing) of nested functions, but we wouldn't reach that bit of code from `AOTLowerMain` since it doesn't support nested functions? It's not entirely clear to me where this came from. 

My opinion regarding to the testing has not changed - I agree with @mbaret in general that we shouldn't test implementation details, but I'd test `ExprAllocator` in this case since besides being an implementation detail it is a chunky pass with complex logic. However, I don't perceive it as a fundamental issue, so I have no intention to block this PR. I also don't think the "ensure test coverage of the new feature" of the code quality guidelines is particularly helpful here since it is not defined what "test coverage" means. So I'd approve this PR once I'll understand better what is going on with the `CreateStorage` (since clearly I'm missing something).

**Review by Mousius — APPROVED** [review]
I would move the `CreateStorage` function out of the header, but otherwise I appreciate your patience in building up our test coverage @mbaret 😸 Will leave that bit with @ekalda

**Review by ekalda — APPROVED** [review]
Thanks for the clarification @mbaret! I hadn't realised the handling of nested function was there before. I'm happy with the current state of testing of this PR (I suppose you could move `CreateStorage` out of the header the way @Mousius suggests in a future patch). I agree that it would be more appropriate to extend the testing coverage when we make `ExprAllocator` a general component, so it's probably out of scope of this refactor. 

Thanks again for all the effort! :medal_sports:

**Comment by lhutton1** [comment]
Thanks @mbaret @areusch @Mousius @ekalda!
